packages/yew-functional/tests/use_context.rs

```diff
@@ -233,36 +233,37 @@ fn use_context_update_works() {
         fn run(_props: &Self::TProps) -> Html {
             type MyContextProvider = ContextProvider<Rc<MyContext>>;
 
-            let (ctx, set_ctx) = use_state(|| MyContext("hello".into()));
+            let ctx = use_state(|| MyContext("hello".into()));
             let rendered = use_ref(|| 0);
 
             // this is used to force an update specific to test-2
-            let (magic_rc, set_magic) = use_state(|| 0);
+            let magic_rc = use_state(|| 0);
```

Comment:
Just wondering, can we do `use_state(0)` here, let `use_state` support both function and value to be more ergonomic?

Reply:
As far as I know, unfortunately, it is impossible. This would require overloading where one `use_state` implementation takes a value and another one takes in a closure. Rust doesn't support overloading so we can't have those implementations.

This *might* be possible with specialization but that's nightly only and I've never used it so I don't know how to implement it.

Reply:
Couldn't we do something like this?

```rust
trait HookInput<T> {}

impl<T, S: Fn() -> T> HookInput<T> for S {}

impl HookInput<i32> for i32 {}

impl HookInput<u32> for u32 {}

fn use_state<T>(_: impl HookInput<T>) {}

fn main() {
    use_state(0 as u32);
    use_state(0 as i32);
    use_state(|| 0);
}
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f683268400ff1ec69f5451f9c5cb4405)